extensions/hal/runtime/src/main/java/io/quarkus/hal/HalService.java

```diff
@@ -0,0 +1,79 @@
+package io.quarkus.hal;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Service with Hal utilities. This service is used by the Resteasy Links, Resteasy Reactive Links and the
+ * Rest Data Panache extensions.
+ */
+@SuppressWarnings("unused")
+public abstract class HalService {
+
+    private static final String SELF_REF = "self";
+
+    /**
+     * Wrap a collection of objects into a Hal collection wrapper by resolving the Hal links.
+     * The Hal collection wrapper is then serialized by either json or jackson.
+     *
+     * @param collection The collection of objects to wrap.
+     * @param collectionName The name that will include the collection of objects within the `_embedded` Hal object.
+     * @param entityClass The class of the objects in the collection. If null, it will not resolve the links for these objects.
+     * @return The Hal collection wrapper instance.
+     */
+    public HalCollectionWrapper toHalCollectionWrapper(Collection<Object> collection, String collectionName,
+            Class<?> entityClass) {
+        List<HalEntityWrapper> items = collection.stream()
+                .map(this::toHalWrapper)
+                .collect(Collectors.toList());
+
+        Map<String, HalLink> classLinks = Collections.emptyMap();
+        if (entityClass != null) {
+            classLinks = getClassLinks(entityClass);
+        }
+
+        return new HalCollectionWrapper(items, collectionName, classLinks);
+    }
+
+    /**
+     * Wrap an entity into a Hal instance by including the entity itself and the Hal links.
+     *
+     * @param entity The entity to wrap.
+     * @return The Hal entity wrapper.
+     */
+    public HalEntityWrapper toHalWrapper(Object entity) {
+        return new HalEntityWrapper(entity, getInstanceLinks(entity));
+    }
+
+    /**
+     * Get the HREF link with reference `self` from the Hal links of the entity instance.
+     *
+     * @param entity The entity instance where to get the Hal links.
+     * @return the HREF link with rel `self`.
+     */
+    public String getSelfLink(Object entity) {
+        return Optional.ofNullable(getInstanceLinks(entity).get(SELF_REF))
+                .map(HalLink::getHref)
```

Comment:
With Quarkus aiming to provide very small runtime footprint, this is the kind of thing we usually avoid :)

Reply:
Changed to not using optional and map.